### PR TITLE
Master accounts budget newapi ajo

### PR DESCRIPTION
**Pull request by ajo-odoo**
Task: https://www.odoo.com/web#id=12160&view_type=form&model=project.task&menu_id=3942&action=327
Pad: https://pad.odoo.com/p/r.O4uzzrog0cpR2iVG

### CODE REVIEW

File: addons/account_budget/wizard/account_budget_analytic.py [hunk 1]
```diff
@@ -1,31 +1,27 @@
...
-            context = {}
-        data = self.read(cr, uid, ids, context=context)[0]
+    date_from = fields.Date(string='Start of period', required=True, default=datetime(int(fields.Date.today()[:4]), 01, 01))
+    date_to = fields.Date(string='End of period', required=True, default=fields.Date.today)
+
+    @api.multi
+    def check_report(self):
+        self.ensure_one()
+        if self.env.context is None:
+            self.env.context = {}
```
Comment (WHPODOO) [inline addons/account_budget/wizard/account_budget_analytic.py]: Not require if you want remove then remove it : 
```
if self.env.context is None:
    self.env.context = {}
```

File: addons/account_budget/wizard/account_budget_crossovered_report.py [hunk 2]
```diff
@@ -1,32 +1,27 @@
...
-        if context is None:
-            context = {}
-        data = self.read(cr, uid, ids, context=context)[0]
+    date_from = fields.Date(string='Start of period', required=True, default=datetime(int(fields.Date.today()[:4]), 01, 01))
+    date_to = fields.Date(string='End of period', required=True, default=fields.Date.today)
+
+    @api.multi
+    def check_report(self):
+        if self.env.context is None:
+            self.env.context = {}
```
Comment (WHPODOO) [inline addons/account_budget/wizard/account_budget_crossovered_report.py]: Not require if you want remove then remove it : 
```
if self.env.context is None:
    self.env.context = {}
```

File: addons/account_budget/models/crossovered_budget.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+    def budget_done(self):
+        return self.write({'state': 'done'})
+
+
+class CrossoveredBudgetLines(models.Model):
+    _name = "crossovered.budget.lines"
+    _description = "Budget Line"
+
+    crossovered_budget_id = fields.Many2one('crossovered.budget', string='Budget', ondelete='cascade', select=True, required=True)
```
Comment (gpa-odoo) [inline addons/account_budget/models/crossovered_budget.py]: Remove select=True, use index

File: addons/account_budget/models/crossovered_budget.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            if not acc_ids:
+                raise UserError(_("The Budget '%s' has no accounts!") % ustr(line.general_budget_id.name))
+            if line.analytic_account_id.id:
+                analytic_lines = AccountAnalyticLine.search([('account_id', '=', self.analytic_account_id.id), ('date', '>=', self.date_from), ('date', '<=', self.date_to), ('general_account_id', 'in', acc_ids)])
+                for analytic_line in analytic_lines:
+                    amount += analytic_line.amount
+            result[line.id] = amount
+        return result
+
+    def _prac(self):
```
Comment (gpa-odoo) [inline addons/account_budget/models/crossovered_budget.py]: why here @api missing?

File: addons/account_budget/models/crossovered_budget.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+                elif line_timedelta.days > 0 and today < strToDatetime(line.date_to):
+                    # If today is between the budget line date_from and date_to
+                    theo_amt = (elapsed_timedelta.total_seconds() / line_timedelta.total_seconds()) * line.planned_amount
+                else:
+                    theo_amt = line.planned_amount
+
+            res[line.id] = theo_amt
+        return res
+
+    def _theo(self):
```
Comment (gpa-odoo) [inline addons/account_budget/models/crossovered_budget.py]: same as above

File: addons/account_budget/models/crossovered_budget.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+    name = fields.Char(string='Budget Name', required=True, states={'done': [('readonly', True)]})
+    creating_user_id = fields.Many2one('res.users', string='Responsible', default=lambda self: self.env.uid,)
+    date_from = fields.Date(string='Start Date', required=True, states={'done': [('readonly', True)]})
+    date_to = fields.Date(string='End Date', required=True, states={'done': [('readonly', True)]})
+    state = fields.Selection([('draft', 'Draft'), ('cancel', 'Cancelled'), ('confirm', 'Confirmed'), ('validate', 'Validated'), ('done', 'Done')], 'Status', select=True, required=True, readonly=True, copy=False, track_visibility='always', default='draft')
+    crossovered_budget_line = fields.One2many('crossovered.budget.lines', 'crossovered_budget_id', string='Budget Lines', states={'done': [('readonly', True)]}, copy=True)
+    company_id = fields.Many2one('res.company', string='Company', required=True, default=lambda self: self.env['res.company']._company_default_get('account.budget.post'))
+
+    def budget_confirm(self):
```
Comment (gpa-odoo) [inline addons/account_budget/models/crossovered_budget.py]: here not need to give api ?

File: addons/account_budget/tests/test_account_budget_report.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+        self.AccountBudgetPost = self.env['account.budget.post']
+        self.Account = self.env['account.account']
+        self.o_income = self.Account.create({'name': 'Opening Income', 'code': 'x1017', 'user_type_id': self.env['account.account.type'].create({'name': 'Other Income', 'type': 'other'}).id})
+        self.account_budget_post_sale0 = self.AccountBudgetPost.create({'name': 'Test of sale', 'account_ids': [self.o_income.id]})
+
+    def test_00_account_report(self):
+
+        # Print the Analytic Budget Report through wizard
+
+        ctx = {'model': 'account.analytic.account', 'active_ids': self.analytic_internal.id}
```
Comment (gpa-odoo) [inline addons/account_budget/tests/test_account_budget_report.py]: Here in active_ids, you have to pass list of id

File: addons/account_budget/tests/test_account_budget_report.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+        test_reports.try_report_action(self.env.cr, self.env.uid, 'action_account_budget_analytic', wiz_data={}, context=ctx, our_module='account_budget')
+
+        # Print the Budget Report through wizard
+
+        ctx = {'model': 'account.budget.post', 'active_ids': self.account_budget_post_sale0.id}
```
Comment (gpa-odoo) [inline addons/account_budget/tests/test_account_budget_report.py]: same as above

File: addons/account_budget/tests/test_theoreticalamount.py [hunk 9]
```diff
@@ -12,90 +10,88 @@
...
             'name': 'test budget name',
             'date_from': '2014-01-01',
             'date_to': '2014-12-31',
         })
-        crossovered_budget_line_obj = self.registry('crossovered.budget.lines')
         user_type_id = self.ref('account.data_account_type_revenue')
         tag_id = self.ref('account.account_tag_operating')
-        account_rev_id = self.registry('account.account').create(cr, uid, {'code': 'X2020', 'name': 'Budget - Test Revenue Account', 'user_type_id': user_type_id, 'tag_ids': [(4, tag_id, 0)]})
-        buget_post = self.registry('account.budget.post').create(cr, uid, {
+        account_rev_id = self.env['account.account'].create({'code': 'X2020', 'name': 'Budget - Test Revenue Account', 'user_type_id': user_type_id, 'tag_ids': [(4, tag_id, 0)]})
```
Comment (gpa-odoo) [inline addons/account_budget/tests/test_theoreticalamount.py]: here account_rev_id is record set

File: addons/account_budget/tests/test_theoreticalamount.py [hunk 10]
```diff
@@ -12,90 +10,88 @@
 class TestTheoreticalAmount(TransactionCase):
     def setUp(self):
         super(TestTheoreticalAmount, self).setUp()
-        cr, uid, = self.cr, self.uid
-        crossovered_budget_id = self.registry('crossovered.budget').create(cr, uid, {
+        crossovered_budget_id = self.env['crossovered.budget'].create({
```
Comment (gpa-odoo) [inline addons/account_budget/tests/test_theoreticalamount.py]: crossovered_budget_id is record set not id, same thing check other

File: addons/account_budget/wizard/account_budget_analytic_view.xml [hunk 11]
```diff
@@ -1,7 +1,5 @@
 <?xml version="1.0" encoding="utf-8"?>
-<openerp>
-     <data>
-
+<odoo>
         <record id="account_budget_analytic_view" model="ir.ui.view">
```
Comment (gpa-odoo) [inline addons/account_budget/wizard/account_budget_analytic_view.xml]: check indentation, same other files to

File: addons/account_budget/wizard/account_budget_analytic.py [hunk 12]
```diff
@@ -1,31 +1,25 @@
...
+        data = {
+            'ids': self.env.context.get('active_ids', []),
             'model': 'account.analytic.account',
-            'form': data
+            'form': self.read()[0]
         }
-        datas['form']['ids'] = datas['ids']
-        return self.pool['report'].get_action(cr, uid, [], 'account_budget.report_analyticaccountbudget', data=datas, context=context)
+        data['form']['ids'] = data['ids']
+        act_id = self.env['account.analytic.account'].browse(self.env.context.get('active_ids'))
```
Comment (gpa-odoo) [inline addons/account_budget/wizard/account_budget_analytic.py]: improve variable name its record set

File: addons/account_budget/wizard/account_budget_crossovered_report.py [hunk 13]
```diff
@@ -1,32 +1,27 @@
...
-            'form': data
+            'form': self.read()[0]
         }
-        datas['form']['ids'] = datas['ids']
-        datas['form']['report'] = 'analytic-full'
-        return self.pool['report'].get_action(cr, uid, [], 'account_budget.report_crossoveredbudget', data=datas, context=context)
+        data['form']['ids'] = data['ids']
+        data['form']['report'] = 'analytic-full'
+        act_id = self.env['crossovered.budget'].browse(self.env.context.get('active_ids'))
+        return self.env['report'].get_action(act_id, 'account_budget.report_crossoveredbudget', data=data)
```
Comment (gpa-odoo) [inline addons/account_budget/wizard/account_budget_crossovered_report.py]: same as above

File: addons/account_budget/report/analytic_account_budget_report.py [hunk 14]
```diff
@@ -140,17 +114,27 @@ def funct_total(self, form):
         result = []
         res = {}
```
Comment (gpa-odoo) [inline addons/account_budget/report/analytic_account_budget_report.py]: remove res unused

File: addons/account_budget/__openerp__.py [hunk 15]
```diff
@@ -33,25 +33,20 @@
...
+    'demo': ['data/account_budget_demo.xml', 'data/account_budget_demo.yml'],
     'test': [
         '../account/test/account_minimal_test.xml',
-        'account_budget_demo.yml',
-        'test/account_budget.yml',
-        'test/account_budget_report.yml',
     ],
-    'installable': True,
-    'auto_install': False,
+
```
Comment (gpa-odoo) [inline addons/account_budget/__openerp__.py]: Remove version!

File: addons/account_budget/models/crossovered_budget.py [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+                if elapsed_timedelta.days < 0:
+                    # If the budget line has not started yet, theoretical amount should be zero
+                    theo_amt = 0.00
+                elif line_timedelta.days > 0 and today < strToDatetime(line.date_to):
+                    # If today is between the budget line date_from and date_to
+                    theo_amt = (elapsed_timedelta.total_seconds() / line_timedelta.total_seconds()) * line.planned_amount
+                else:
+                    theo_amt = line.planned_amount
+
+            res[line.id] = theo_amt
```
Comment (gpa-odoo) [inline addons/account_budget/models/crossovered_budget.py]: Is "res" used now?

File: addons/account_budget/tests/test_account_budget.py [hunk 17]
```diff
@@ -0,0 +1,49 @@
+# -*- coding: utf-8 -*-
+
+from datetime import datetime
+from odoo import fields
+from openerp.addons.account_budget.tests.test_common import TestAccountBudgetCommon
```
Comment (gpa-odoo) [inline addons/account_budget/tests/test_account_budget.py]: group imports!

**Comment by hmo-odoo** [comment]
continue work on this PR: https://github.com/odoo/odoo/pull/11567

**Pull request closed without merging**